File: networkclient/src/main/java/cm/aptoide/pt/networkclient/okhttp/UserAgentInterceptor.java
package cm.aptoide.pt.networkclient.okhttp;

import android.text.TextUtils;
import cm.aptoide.pt.v8engine.crashreports.CrashReport;
import java.io.IOException;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class UserAgentInterceptor implements Interceptor {

  private final UserAgentGenerator userAgentGenerator;

  public UserAgentInterceptor(UserAgentGenerator userAgentGenerator) {
    this.userAgentGenerator = userAgentGenerator;
  }

  @Override public Response intercept(Chain chain) throws IOException {
    Request originalRequest = chain.request();

    String userAgent = null;
    try {
      userAgent = userAgentGenerator.generateUserAgent();
    } catch (Exception e) {
      CrashReport.getInstance().log(e);
    }

    Response response;
    try {
      if (!TextUtils.isEmpty(userAgent)) {
        Request requestWithUserAgent =
            originalRequest.newBuilder().header("User-Agent", userAgent).build();
        response = chain.proceed(requestWithUserAgent);
      } else {
        response = chain.proceed(originalRequest);
      }
      return response;
    } catch (IOException e) {
      // something bad happened if we reached here.
      CrashReport.getInstance().log(e);
      throw e;
    }
  }
}

File: payments/base/network/src/main/java/com/appcoins/payments/network/UserAgentInterceptor.kt
package com.appcoins.payments.network

import okhttp3.Interceptor
import okhttp3.Response
import javax.inject.Inject

class UserAgentInterceptor @Inject constructor(private val provider: GetUserAgent) :
  Interceptor {

  override fun intercept(chain: Interceptor.Chain): Response {
    val userAgent = provider()
    val newRequest = chain.request().newBuilder().header("User-Agent", userAgent).build()
    return chain.proceed(newRequest)
  }
}

interface GetUserAgent {
  operator fun invoke(): String
}
